keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,821 @@
+  - However, enabling this feature will ensure that the correct resource information is pushed across the cluster.
+
+## Design Details
+
+Below diagram is shows the interaction between kubelet, node and cAdvisor.
+
+
+```mermaid
+sequenceDiagram
+    participant node
+    participant kubelet
+    participant cAdvisor-cache
+    participant machine-info
+    kubelet->>cAdvisor-cache: Poll
+    cAdvisor-cache->>machine-info: fetch
+    machine-info->>cAdvisor-cache: update
+    cAdvisor-cache->>kubelet: update
+    kubelet->>node: node status update
+    kubelet->>node: re-initialize resource managers
+    kubelet->>node: Recalculate and update OOMScoreAdj <br> and Swap limit of pods
+```
+
+The interaction sequence is as follows
+1. Kubelet will be polling in interval to fetch the machine resource information from cAdvisor's cache, Which is currently updated every 5 minutes.
+2. Kubelet's cache will be updated with the latest machine resource information.
+3. Node status updater will update the node's status with the latest resource information.
+4. Kubelet will reinitialize the resource managers to keep them up to date with dynamic resource changes.
+5. Kubelet will recalculate and update the OOMScoreAdj and swap limit for the existing pods.
+
+With increase in cluster resources the following components will be updated
+
+1. Scheduler
+   * Scheduler will automatically schedule any pending pods.
+
+2. Update in Node allocatable capacity.
+
+3. Resource managers will re-initialised.
+
+4. Change in OOM score adjust
+    * Currently, the OOM score adjust is calculated by
+      `1000 - (1000*containerMemReq)/memoryCapacity`
+    * Increase in memoryCapacity will result in updated OOM score adjust for pods deployed post resize and also recalculate the same for existing pods.
+
+5. Change in Swap Memory limit
+   * Currently, the swap memory limit is calculated by 
+ `(<containerMemoryRequest>/<nodeTotalMemory>)*<totalPodsSwapAvailable>`
+   * Increase in nodeTotalMemory or totalPodsSwapAvailable will result in updated swap memory limit for pods deployed post resize and also recalculate the same for existing pods.
+
+**Proposed Code changes**
+
+**Pseudocode for Resource Hotplug**
+
+```go
+	if utilfeature.DefaultFeatureGate.Enabled(features.NodeResourceHotPlug) {
+		// Handle the node dynamic scale up
+		machineInfo, err := kl.cadvisor.MachineInfo()
+		if err != nil {
+			klog.ErrorS(err, "Error fetching machine info")
+		} else {
+			cachedMachineInfo, _ := kl.GetCachedMachineInfo()
+            // Avoid collector collects it as a timestamped metric
+            // See PR #95210 and #97006 for more details.
+                machineInfo.Timestamp = time.Time{}
+			if !reflect.DeepEqual(cachedMachineInfo, machineInfo) {
+				kl.setCachedMachineInfo(machineInfo)
+
+				// Resync the resource managers
+				if err := kl.containerManager.ResyncComponents(machineInfo); err != nil {
```

Comment:
You probably have captured this in the actual code, but I think the kubelet runtime manager needs to be synced in a similar fashion. Should we capture this in the KEP?

Reply:
Yeah, I have tried something like [this](https://github.com/Karthik-K-N/kubernetes/blob/hot-plug/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go#L138-L140) in PoC, May be I will update the KEP to mention the same. Thanks